modules/optflow/src/rlof/plk_invoker.hpp

```diff
@@ -286,177 +200,130 @@ class TrackerInvoker : public cv::ParallelLoopBody
                     A22 = 0;
                 }
-#ifdef RLOF_SSE
-                qw0 = _mm_set1_epi32(iw00 + (iw01 << 16));
-                qw1 = _mm_set1_epi32(iw10 + (iw11 << 16));
-                __m128 qb0 = _mm_setzero_ps(), qb1 = _mm_setzero_ps(), qb2 = _mm_setzero_ps();
-                __m128 qb3 = _mm_setzero_ps();
-                __m128 mmSumW1 = _mm_setzero_ps(), mmSumW2 = _mm_setzero_ps();
-                __m128 mmSumI = _mm_setzero_ps(), mmSumW = _mm_setzero_ps(), mmSumDI = _mm_setzero_ps();
-                __m128 mmSumIy = _mm_setzero_ps(),  mmSumIx = _mm_setzero_ps();
-                __m128 mmAxx = _mm_setzero_ps(), mmAxy = _mm_setzero_ps(), mmAyy = _mm_setzero_ps();
+#ifdef CV_SIMD128
+                v_int16x8 vqw0 = v_int16x8((short)(iw00), (short)(iw01), (short)(iw00), (short)(iw01), (short)(iw00), (short)(iw01), (short)(iw00), (short)(iw01));
+                v_int16x8 vqw1 = v_int16x8((short)(iw10), (short)(iw11), (short)(iw10), (short)(iw11), (short)(iw10), (short)(iw11), (short)(iw10), (short)(iw11));
+                v_float32x4 vqb0 = v_setzero_f32(), vqb1 = v_setzero_f32(), vqb2 = v_setzero_f32(), vqb3 = v_setzero_f32();
+                v_float32x4 vsumW1 = v_setzero_f32(), vsumW2 = v_setzero_f32();
+                v_float32x4 vsumIy = v_setzero_f32(), vsumIx = v_setzero_f32(), vsumI = v_setzero_f32(), vsumDI = v_setzero_f32();
+                v_float32x4 vAxx = v_setzero_f32(), vAxy = v_setzero_f32(), vAyy = v_setzero_f32();
+                v_int32x4 vdelta = v_setall_s32(1 << (W_BITS1 - 5 - 1));
+                v_int16x8 vmax_val_16 = v_setall_s16(std::numeric_limits<unsigned short>::max());
                 float gainVal = gainVec.x > 0 ? gainVec.x : -gainVec.x;
                 int bitShift = gainVec.x == 0 ? 1 : cvCeil(log(200.f / gainVal) / log(2.f));
-                __m128i mmGainValue_epi16 = _mm_set1_epi16(static_cast<short>(gainVec.x * (float)(1 << bitShift)));
-                __m128i mmConstValue_epi16 = _mm_set1_epi16(static_cast<short>(gainVec.y));
+                v_int16x8 vgain_value = v_setall_s16(static_cast<short>(gainVec.x * (float)(1 << bitShift)));
+                v_int16x8 vconst_value = v_setall_s16(static_cast<short>(gainVec.y));
 #endif
-                for( y = 0; y < winSize.height; y++ )
+                for(int y = 0; y < winSize.height; y++ )
                 {
                     const uchar* Jptr = J.ptr<uchar>(y + inextPt.y, inextPt.x*cn);
                     const uchar* Jptr1 = J.ptr<uchar>(y + inextPt.y + 1, inextPt.x*cn);
                     const short* Iptr  = IWinBuf.ptr<short>(y, 0);
                     const short* dIptr = derivIWinBuf.ptr<short>(y, 0);
                     const tMaskType* maskPtr = winMaskMat.ptr<tMaskType>(y, 0);
-                    x = 0;
-#ifdef RLOF_SSE
-
-                    for( ; x <= winSize.width*cn; x += 8, dIptr += 8*2 )
+#ifdef CV_SIMD128
+                    for(int x = 0 ; x <= winSize.width*cn; x += 8, dIptr += 8*2 )
```

Comment:
Does this loop iterates correctly?
- processed all elements (step is 8). Do we have a "tail"?
- it doesn't access elements outside of allocated buffers. Condition should look like:
    ```
    x <= fullWidth - elementsPerIteration
    ```

There is full [generic code pattern](https://github.com/opencv/opencv/blob/b0563ce368315de482849f2a7575ec6d54dcb793/modules/core/src/mathfuncs.cpp#L611-L626) which using SIMD or other unrolling approaches.

Reply:
There will be a tail, e.g. if winSize.width = 9. When I change `x <= winSize.width*cn` to `x <= winSize.width*cn-8` some elements are not taken into account. In order to avoid access elements outside the winBuffer. This buffer has been allocated with a width computed by `int winMaskwidth = roundUp(winSize.width, 16);` (e.g. see line 72 plk_invoker.hpp)